feat(core): support APP_INITIALIZER work with observable

vthinkxie opened the pull request:
close #15088

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] <del>Docs have been added / updated (for bug fixes / features)</del>


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
feat(core): support APP_INITIALIZER work with observable

This commit adds support for Observables that now can be used as a part of APP_INITIALIZER. Previously, only
Primises were supported.

Closes #15088.

vthinkxie commented:
cc @alxhub

JiaLiPassion reviewed (commented):
since `toPromise` is being deprecated, maybe we need to use something like `await lastValueFrom`?

JiaLiPassion commented:
We also need to update the doc here https://angular.io/guide/dependency-injection-providers

AndrewKushnir commented:
Hi @vthinkxie, sorry we didn't have a chance to review your PR earlier.

As Jia mentioned above, the `toPromise` function is deprecated in RxJS and the `lastValueFrom` function should be used instead. Could you please update the code to use `lastValueFrom` function and also rebase this PR and resolve merge conflicts? Once the PR is updated we'll review the changes and provide additional feedback.

Thank you.

vthinkxie commented:
> Hi @vthinkxie, sorry we didn't have a chance to review your PR sooner.
> 
> As Jia mentioned above, the `toPromise` function is deprecated in RxJS and the `lastValueFrom` function should be used instead. Could you please update the code to use `lastValueFrom` function and also rebase this PR and resolve merge conflicts? Once the PR is updated we'll review the changes and provide additional feedback.
> 
> Thank you.

Hi @JiaLiPassion @AndrewKushnir 
thanks a lot for your review
I have rebased onto master
It seems that `lastValueFrom` is not available in rxjs 6 yet, maybe we can have another PR when rxjs 7 released?

AndrewKushnir commented on packages/core/src/application_init.ts:
```diff
@@ -68,6 +68,8 @@ export class ApplicationInitStatus {
         if (isPromise(initResult)) {
           asyncInitPromises.push(initResult);
+        } else if (isObservable(initResult)) {
+          asyncInitPromises.push(initResult.toPromise());
```
> It seems that lastValueFrom is not available in rxjs 6 yet, maybe we can have another PR when rxjs 7 released?

Makes sense, thanks @vthinkxie. I think we should add a quick comment above that line to provide more context for the followup refactoring (once we update RxJS to v7):

```suggestion
          // Note: `.toPromise()` will be deprecated in RxJS v7 and removed completely in later versions.
          // Once RxJS is updated in the repo to v7, this code should be refactored and the `lastValueFrom` function 
          // should be used instead.
          asyncInitPromises.push(initResult.toPromise());
```

We have [an item on the roadmap](https://angular.io/guide/roadmap#ensure-smooth-adoption-for-future-rxjs-changes-v7-and-beyond) to update RxJS to v7 and if it happens before next major release (v12), we could update that code too. Otherwise, we'll update it once RxJS is updated and the `lastValueFrom` function becomes available.

AndrewKushnir replied:
@vthinkxie it looks like the comment was reverted, could you please add it back, so there is some context for future refactoring? Thank you.

vthinkxie replied:
sure, wait one moment

> @vthinkxie it looks like the comment was reverted, could you please add it back, so there is some context for future refactoring? Thank you.

sure, wait one moment

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -66,6 +67,51 @@ import {inject, TestBed, waitForAsync} from '../testing';
+             expect(completerResolver).toBe(true);
+           });
+         })));
+    });
+    describe('with async observable initializers', () => {
```
Small suggestion to update block title (and also adding an extra line):
```suggestion
    });

    describe('with app initializers represented using observables', () => {
```

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -66,6 +67,51 @@ import {inject, TestBed, waitForAsync} from '../testing';
+        TestBed.configureTestingModule({
+          providers: [
+            {provide: APP_INITIALIZER, multi: true, useValue: () => observable},
+            {
+              provide: APP_INITIALIZER,
+              multi: true,
+              useFactory: initializerFactory,
+              deps: [Injector]
+            },
```
I'm a bit confused why we need 2nd `APP_INITIALIZER` token. Can we refactor/simplify the code to use just one?

vthinkxie replied:
Hi, these lines are following https://github.com/angular/angular/blob/d067dc0cb9a3088a7c3443d33d35f7a0856b8a49/packages/core/test/application_init_spec.ts#L47

I have tried refactor the codes, but it seems that we can not have useFactory and useValue in the same provider in the current angular version, any good idea?

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -66,6 +67,51 @@ import {inject, TestBed, waitForAsync} from '../testing';
+        let initializerFactory = (injector: Injector) => {
+          return () => {
+            const initStatus = injector.get(ApplicationInitStatus);
+            initStatus.donePromise.then(() => {
+              expect(completerResolver).toBe(true);
+            });
```
I believe this can be moved inside the test itself, it looks like a more expected place for the `expect` function call.

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -66,6 +67,51 @@ import {inject, TestBed, waitForAsync} from '../testing';
+           (status as any).runInitializers();
+
+           setTimeout(() => {
+             completerResolver = true;
+             resolve.complete();
```
I'd propose to add a couple more tests to check behavior in the following cases:
- observable completes with an error
- observable emits multiple values and then completes

AndrewKushnir reviewed (changes requested):
@vthinkxie thanks for updating this PR 👍

I've left a few comments, please have a look when you get a chance.

There is a couple additional comments/thoughts:

- I believe we should also update `APP_INITIALIZER` token type to be more precise with respect to the types of expected values that the code handles (Promise, Observable). Note: this is definitely **outside** of the scope of this PR, so we can do that later (this would be a breaking change).  I've created a ticket on that: https://github.com/angular/angular/issues/40729.
- As a part of this PR we should update the `APP_INITIALIZER` docs page ([see here](https://angular.io/api/core/APP_INITIALIZER)) to mention `Observable`s support. 

I've also created a separate ticket to add usage examples for the `APP_INITIALIZER` token in the followup PRs: https://github.com/angular/angular/issues/40730.

Thank you.

vthinkxie commented:
Hi @AndrewKushnir thanks for your comments
I have updated the PR, could you have a look again? thanks!

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,55 @@ import {inject, TestBed, waitForAsync} from '../testing';
     });
+
+    describe('with app initializers represented using observables', () => {
+      let resolve: Subscriber<any>;
```
May be rename `resolve` to `subscriber` (so it's more inline with RxJS terminology)?
```suggestion
      let subscriber: Subscriber<any>;
```

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,55 @@ import {inject, TestBed, waitForAsync} from '../testing';
+
+      it('should update the status once all async observable initializers are done',
+         waitForAsync(inject([ApplicationInitStatus], (status: ApplicationInitStatus) => {
+           (status as any).runInitializers();
```
I'd propose to add a quick comment on why we use `as any` here (and in the second test as well):
```suggestion
           // Accessing internal `runInitializers` function of the `ApplicationInitStatus` class instance
           // for testing purposes to invoke initializer functions.
           (status as any).runInitializers();
```

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,55 @@ import {inject, TestBed, waitForAsync} from '../testing';
+             resolve.complete();
+           });
+
+           resolve.next();
```
Should we invoke this inside the `setTimeout` callback function (not at the top level of this function)?

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,55 @@ import {inject, TestBed, waitForAsync} from '../testing';
+           });
+         })));
+
+      it('should update the status once async observable initializers has error',
```
FYI, it looks like this test is failing on CI in Firefox, see: https://app.circleci.com/pipelines/github/angular/angular/28035/workflows/b6a0fabc-6b3c-44e9-a21d-58691618f33d/jobs/912572
Search for the "should update the status once async observable initializers has error" string in the output.

vthinkxie replied:
Hi, I had fixed the failing test

AndrewKushnir reviewed (commented):
Thanks @vthinkxie, I've left a few more comments.

I think we should also update the `APP_INITIALIZER` token API docs ([see here](https://github.com/angular/angular/blob/master/packages/core/src/application_init.ts#L14-L31)) to mention that it now supports Observables.

vthinkxie commented:
> Thanks @vthinkxie, I've left a few more comments.
> 
> I think we should also update the `APP_INITIALIZER` token API docs ([see here](https://github.com/angular/angular/blob/master/packages/core/src/application_init.ts#L14-L31)) to mention that it now supports Observables.

Hi, I had updated the commits, thanks for your review

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,58 @@ import {inject, TestBed, waitForAsync} from '../testing';
+    describe('with app initializers represented using observables', () => {
+      let subscriber: Subscriber<any>;
+      let observable: Observable<any>;
+      let completerResolver = false;
```
The `completerResolver` is used in existing Promise-based tests, but I think we can rename it in newly-added tests to be a bit more clear on what this flag stands for, for example:
```suggestion
      let initFnInvoked = false;
```

AndrewKushnir replied:
I'd vote to change this in existing tests in the same file for consistency as well.

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,58 @@ import {inject, TestBed, waitForAsync} from '../testing';
+           });
+
+           expect(status.done).toBe(false);
+           status.donePromise.catch(() => {
+             expect(status.done).toBe(false);
+             expect(completerResolver).toBe(true);
+           });
```
I'd propose adding `then` here to make sure it's not invoked when observable emits an error:
```suggestion
           status.donePromise
             .then(() => fail('`donePromise.then` should not be invoked when observable emits an error'))
             .catch(() => {
               expect(status.done).toBe(false);
               expect(completerResolver).toBe(true);
             });
```

AndrewKushnir commented on packages/core/test/application_init_spec.ts:
```diff
@@ -73,5 +74,58 @@ import {inject, TestBed, waitForAsync} from '../testing';
+         })));
+
+      it('should update the status once async observable initializers has error',
+         waitForAsync(inject([ApplicationInitStatus], (status: ApplicationInitStatus) => {
```
I think we should add similar test for Promise-based case to make sure the behavior is the same when Promise is rejected. Could you please create one? I believe it would be similar to the `should update the status once all async promise initializers are done` test above.

AndrewKushnir reviewed (commented):
Thanks for additional changes @vthinkxie! I've left a bit more feedback mostly regarding tests.

I also noticed that one of the CI jobs is failing ([see output here](https://app.circleci.com/pipelines/github/angular/angular/28172/workflows/51493c10-85af-47ce-928b-c9d4620229f6/jobs/914818)). The problem appeared because a new `isObservable` symbol is now added into a couple bundles for the test apps. This is expected and you can update golden files by running the following commands:

```
yarn bazel run --config=ivy //packages/core/test/bundling/forms:symbol_test.accept
yarn bazel run --config=ivy //packages/core/test/bundling/router:symbol_test.accept
```

Running these commands will update a couple files in the `goldens` folder, please also commit these changes (as a fixup commit to the main one).

Please let me know if you have any questions or any assistance is required.

Thank you.

vthinkxie commented:
> Thanks for additional changes @vthinkxie! I've left a bit more feedback mostly regarding tests.
> 
> I also noticed that one of the CI jobs is failing ([see output here](https://app.circleci.com/pipelines/github/angular/angular/28172/workflows/51493c10-85af-47ce-928b-c9d4620229f6/jobs/914818)). The problem appeared because a new `isObservable` symbol is now added into a couple bundles for the test apps. This is expected and you can update golden files by running the following commands:
> 
> ```
> yarn bazel run --config=ivy //packages/core/test/bundling/forms:symbol_test.accept
> yarn bazel run --config=ivy //packages/core/test/bundling/router:symbol_test.accept
> ```
> 
> Running these commands will update a couple files in the `goldens` folder, please also commit these changes (as a fixup commit to the main one).
> 
> Please let me know if you have any questions or any assistance is required.
> 
> Thank you.

fixed, thanks so much for your assistance!

AndrewKushnir reviewed (approved):
The changes look good to me, thanks @vthinkxie 👍 

I'm adding @JoostK as a reviewer to this PR to get his feedback too.

JoostK commented on packages/core/test/application_init_spec.ts:
```diff
@@ -28,50 +28,118 @@ import {inject, TestBed, waitForAsync} from '../testing';
+        });
+      });
+
+      it('should update the status once all async observable initializers are done',
```
```suggestion
      it('should update the status once all async observable initializers are completed',
```

JoostK commented on packages/core/test/application_init_spec.ts:
```diff
@@ -28,50 +28,118 @@ import {inject, TestBed, waitForAsync} from '../testing';
+
+           setTimeout(() => {
+             initFnInvoked = true;
+             subscriber.next();
```
Could you add another case where this `next` call is omitted; it's only the completion signal that should be load-bearing, not any of the values. Perhaps it would also be a good idea to add a scenario with an observable that emits multiple values and then completes.

JoostK replied:
Another interesting case is when the observable was completed synchronously. In my opinion the `ApplicationInitStatus.done` should also become `true` synchronously in that case (if there are no other async completion signals, that is), but the PR in its current state introduces an extra tick before `ApplicationInitStatus.done` becomes true (which could be unexpected for synchronously completed observables)

vthinkxie replied:
Hi, I just added two tests to cover these cases.

JoostK replied:
The case where the observable is completed synchronously has not been covered, though. What I'm suggesting is that it could work as follows:

```typescript
let complete = false;
const promise = new Promise<void>((resolve, reject) => {
  initResult.subscribe({
    error: reject,
    complete() {
      complete = true;
      resolve();
    },
  });
});
if (!complete) {
  asyncInitPromises.push(promise);
}
```

Whereas currently it behaves like:

```typescript
const promise = new Promise<void>((resolve, reject) => {
  initResult.subscribe({
    error: reject,
    complete: resolve,
  });
});
asyncInitPromises.push(promise);
```

The difference is that the first example identifies when the observable completes immediately, e.g. in cases `EMPTY` is returned, in which case the promise is _not_ queued.

I don't have a strong feeling about either approach, although I consider the first example to be more correct. @AndrewKushnir  do you have any preference here? Whichever approach is decided upon should be captured in a test (where `subscriber.complete()` is called before `runInitializers()`).

JoostK replied:
Andrew and I chatted a bit about this and decided that always queuing the promise is fine; so the production code can stay as is (we'll deal with refactoring `toPromise` into a viable alternative once RxJS 7 is targeted). What remains to be done is to capture this scenario in a test, i.e. run `subscriber.complete()` _before_ `runInitializers`. @vthinkxie could you please add such a test?

vthinkxie replied:
Thanks so much for pointing it out @JoostK 
I have updated the test case
https://github.com/angular/angular/pull/33222/files#diff-f3cc58ddd1ff9fbd2eb6554c40f7fc2892e09120a66493e8b93909a7c882c2d8R142
could you have a look?

JoostK commented on packages/core/src/application_init.ts:
```diff
@@ -68,6 +68,11 @@ export class ApplicationInitStatus {
         if (isPromise(initResult)) {
           asyncInitPromises.push(initResult);
+        } else if (isObservable(initResult)) {
+          // Note: `.toPromise()` will be deprecated in RxJS v7 and removed completely in later
+          // versions. Once RxJS is updated in the repo to v7, this code should be refactored and
+          // the `lastValueFrom` function should be used instead.
```
I thought this would be the case, but `lastValueFrom` is fundamentally different to `toPromise` w.r.t. its behavior for empty observables: it throws an `EmptyError` instead of returning `undefined`. For this use-case I believe it's desirable to also accept observables that only complete without ever having emitted a value, so `lastValueFrom` would not be correct here.

vthinkxie replied:
thanks, I have removed this comment here.

vthinkxie commented:
Hi @JoostK
I just added two tests to cover these cases.

1. Another case where this next call is omitted
2. ApplicationInitStatus.done would become true in the promise.then (omit the setTimeout)

could you have a look, thanks

AndrewKushnir commented:
Thanks for the update @vthinkxie. Could you please have a look at [this comment](https://github.com/angular/angular/pull/33222/files#r571347224) as well?

vthinkxie commented:
> Thanks for the update @vthinkxie. Could you please have a look at [this comment](https://github.com/angular/angular/pull/33222/files#r571347224) as well?

Hi, I removed the comments following @JoostK 's suggestion, not sure should do it or not.
I just move it back now.

https://github.com/angular/angular/pull/33222/files#r573676610

AndrewKushnir commented:
Thanks for the reply @vthinkxie. Sorry I must have missed the comment from @JoostK.

> lastValueFrom is fundamentally different to toPromise w.r.t. its behavior for empty observables: it throws an EmptyError instead of returning undefined. For this use-case I believe it's desirable to also accept observables that only complete without ever having emitted a value, so lastValueFrom would not be correct here.

Based on this comment it looks like adding the `toPromise` call now would increase the technical debt later, so we should consider handing this case without relying on `toPromise` from the very beginning (update the changes in this PR to do that). @JoostK what do you think?

JoostK commented:
> Based on this comment it looks like adding the `toPromise` call now would increase the technical debt later, so we should consider handing this case without relying on `toPromise` from the very beginning (update the changes in this PR to do that). @JoostK what do you think?

I don't have a strong preference either way; but indeed we can avoid the need to migrate away from `toPromise` if we never actually start using it (this wasn't the case for `lastValueFrom` as that's not yet available to us today.). So that would mean actually changing the `toPromise` into:

```ts
asyncInitPromises.push(new Promise<void>((resolve, reject) => obs.subscribe({ complete: resolve, error: reject })));
```

(or something like that)

AndrewKushnir commented:
Thanks for the reply @JoostK.

@vthinkxie I think it makes sense to avoid `toPromise` and use the solution proposed by Joost above (since we won't be able to replace it with another function from RxJS). Could you please update the code when you get a chance?

Thank you.

AndrewKushnir commented on packages/core/src/application_init.ts:
```diff
@@ -68,6 +68,9 @@ export class ApplicationInitStatus {
         if (isPromise(initResult)) {
           asyncInitPromises.push(initResult);
+        } else if (isObservable(initResult)) {
+          asyncInitPromises.push(new Promise<void>(
+              (resolve, reject) => initResult.subscribe({complete: resolve, error: reject})));
```
Minor suggestion (for better readability):
```suggestion
          const observableAsPromise = new Promise<void>((resolve, reject) => {
            initResult.subscribe({complete: resolve, error: reject});
          });
          asyncInitPromises.push(observableAsPromise);
```

vthinkxie commented:
Hi, @AndrewKushnir and @JoostK 
I had updated the code, thanks for your suggestion

AndrewKushnir reviewed (approved):
@vthinkxie thanks for addressing all the feedback in this PR 👍 

The code looks good to me and we'll need to get an approval from @JoostK as well.

The next step (after final approval) would be to run tests in Google's codebase (we'll take care of that) and once it's done, we'll be able to add this PR to the merge queue.

Thank you.

JoostK commented on packages/core/test/application_init_spec.ts:
```diff
@@ -28,50 +28,152 @@ import {inject, TestBed, waitForAsync} from '../testing';
+           });
+         })));
+
+      it('should update the status if all async observable initializers are completed before runInitializers',
```
I'd propose to change this into:

```suggestion
      it('should update the status if all async observable initializers are completed synchronously upon subscription',
```

The ordering with respect to `runInitializers` is an implementation detail of the test to achieve synchronous completion, by ensuring the Observable immediately completes.

AndrewKushnir replied:
@vthinkxie just a quick update: we've discussed this with Joost and I plan to push a commit to this PR with tests refactoring that would test `ApplicationInitStatus` directly to avoid TestBed side-effects of invoking `runInitializers`. I will leave a comment here once it's done. Thank you.

AndrewKushnir committed:
test(core): refactor ApplicationInitStatus tests to avoid TestBed side-effects

Currently TestBed (both ViewEngine and Ivy) invoke `ApplicationInitStatus.runInitializers` as a part of the
bootstrap process to mimic real bootstrap steps. This is problematic for the `ApplicationInitStatus` class
tests since the `runInitializers` call performed by TestBed interfere with actual tests.

This commit updates ApplicationInitStatus tests to interact with the class directly instead of relying on TestBed
APIs to retrieve the class though DI.

AndrewKushnir commented:
Quick update: I've pushed an additional commit (https://github.com/angular/angular/pull/33222/commits/26ce63f3e61d6a431eb65854d2ed01c310feef24) (and squashed existing ones) as per  https://github.com/angular/angular/pull/33222#discussion_r578804087.

@vthinkxie @JoostK could you please have a look at the changes altogether to see if anything is missing?

Thank you.

JoostK committed:
fixup! test(core): refactor ApplicationInitStatus tests to avoid TestBed side-effects

AndrewKushnir commented:
[Presubmit](http://test/OCL:358462716:BASE:358462743:1613767339101:ee998010).

AndrewKushnir commented:
@vthinkxie just want to let you know that tests in Google's codebase went well, the PR was approved by all reviewers and it's ready for merge, so I'm adding it to the merge queue.

Thanks for implementing this feature and addressing all the feedback 👍

vthinkxie commented:
Hi @AndrewKushnir and @JoostK 

thanks a lot for your review, looking forward to have this feature in the next angular version!

loxy commented:
Finally it came to an end! Nearly...
Now I know why it took so much time to include this feature to Angular. It was a pleasure to follow your work! Thank you all guys for supporting such a great tool!!! ❤️ Angular

SandraVanDijk commented:
@vthinkxie Could you please tell me how to use this functionality? 

I now have this: 
      provide: APP_INITIALIZER,
      useFactory: (authService: AuthService) => () => { return authService.retrieveToken(); },
      deps: [AuthService],
      multi: true

I have angular version 12.0.0-next.5 installed, but it does not wait until the observable is resolved. The function authService.retrieveToken() does a http get call.

Hope you can give a bit more explanation. Thanks!

AndrewKushnir commented:
Hi @SandraVanDijk, could you please create a minimum repro using [Stackblitz](https://stackblitz.com/), so that we can have a look? Note: you could update package versions in the `package.json` file there (to `12.0.0-next.5`) and the necessary packages would be installed. Once the repro is available, please create a new ticket via https://github.com/angular/angular/issues. Thank you.

SandraVanDijk commented:
@AndrewKushnir I tried to make a Stackblitz as requested, and I updated the packages.json as you told me. After I press save it did install packages and then it says "running ngcc", it doing that now for 2 hours. So I can not put an example there, do you have another suggestion on how to share my code?

AndrewKushnir commented:
> Is there maybe another way on how to share my code?

You could also try to create a new app (using `ng new`), update packages in `package.json` file, add relevant code needed for repro  and put it on GitHub as a new repository.

SandraVanDijk commented:
Hi @AndrewKushnir, thanks for that tip! While making a new repository I figured out what my mistake was. The function I was calling did not return an observable. So it made total sense APP_INILIALIZER did not wait. 

I also found today this documentation:  [https://dzhavat.github.io/2021/02/25/using-observable-in-app-initializer.html](url), maybe this helps somebody else. 

Thanks for your help!
